gammapy/scripts/jupyter.py

```diff
@@ -0,0 +1,149 @@
+        for cellnumber, cell in enumerate(nb.cells):
+            fmt = nb.cells[cellnumber]['source']
+            if nb.cells[cellnumber]['cell_type'] == 'code':
+                try:
+                    semicolon = 0
+                    fmt = comment_magics(fmt)
+                    if fmt.endswith(';'):
+                        semicolon = 1
+                    fmt = format_str(src_contents=fmt,
+                                     line_length=79).rstrip()
+                    if semicolon:
+                        fmt += ';'
+                except Exception as ex:
+                    logging.info(ex)
+                fmt = fmt.replace("###-MAGIC COMMAND-", "")
+            nb.cells[cellnumber]['source'] = fmt
+
+        # write formatted notebook
+        nbformat.write(nb, str(notebook))
+
+        # inform
+        print('Jupyter notebook {} painted in black.'.format(str(notebook)))
+
+
+@click.command(name='stripout')
+@click.pass_context
+def cli_jupyter_stripout(ctx):
+    """Strip output cells."""
+
+    nblist = build_nblist(ctx)
+
+    for notebook in nblist:
+
+        try:
+            subprocess.call("nbstripout '{}'".format(notebook), shell=True)
+            print('Jupyter notebook {} stripped out.'.format(str(notebook)))
+        except Exception as ex:
+            log.error('Error stripping file {}'.format(str(notebook)))
+            log.error(ex)
+
+
+@click.command(name='execute')
+@click.pass_context
+def cli_jupyter_execute(ctx):
+    """Execute jupyter notebook."""
+
+    nblist = build_nblist(ctx)
+
+    for notebook in nblist:
+
+        try:
+            t = time.time()
+            subprocess.call(
+                "jupyter nbconvert --allow-errors --ExecutePreprocessor.timeout=None --ExecutePreprocessor.kernel_name=python3 --to notebook --execute '{}'.format(notebook) --inplace",
+                shell=True)
+            t = (time.time() - t) / 60
+            print('Executing duration: {:.2f} mn'.format(t))
+        except Exception as ex:
+            log.error('Error executing file {}'.format(str(notebook)))
+            log.error(ex)
+
+
+@click.command(name='test')
+@click.pass_context
+def cli_jupyter_test(ctx):
+    """Check if jupyter notebook is broken."""
+
+    notebook = Path(ctx.obj['file'])
+    folder = Path(ctx.obj['fold'])
+    ignorelist = []
+
+    if ctx.obj['file']:
+        # ignore all files except jupyterfile
+        for f in notebook.parent.iterdir():
+            if notebook.name != f.name and f.name.endswith('.ipynb'):
+                nbname = f.name.replace('.ipynb', '')
+                ignorelist.append(nbname)
+        folder = notebook.parent
+
+        print('oups')
+
+    testnb = testipynb.TestNotebooks(
+        directory=str(folder), ignore=ignorelist)
+    TestCase.assertTrue(testnb, testnb.run_tests())
+
+
+def comment_magics(input):
+    """Comment magic commands when formatting cells."""
+
+    lines = input.splitlines(True)
+    output = ""
```

Comment:
Again, this would be little nicer as a generator function that yields the lines.

Reply:
Well, we need the whole code in cell to send it to black, not only the lines.
I do not see how to a generator may help..

Reply:
The generator helps a little bit, it reduces the number of lines you have here, and it's also a bit more efficient. Note that Python strings are immutable, so doing "+" repeatedly to build larger and larger strings isn't efficient.
If you yield strings for each line here, then in the caller you can do this:
```
txt = '\n'.join(comment_magics(input))
```
I think str.join will accept a generator that yields strings.

It's not a big improvement, just a very minor one.

Reply:
Ok, I got it 👍 
Thanks for the review